File: python-templates/seg-tree.py
class SegTree:
    def __init__(self, arr) -> None:
        self.arr = arr
        self.t = [0 for _ in range(4 * len(arr))]

    def merge(self, x, y):
        return x + y

    def build(self, v, tl, tr) -> None:
        if tr == tl:
            self.t[v] = self.arr[tl]
            return
        tm = (tl + tr) // 2
        self.build(2 * v, tl, tm)
        self.build(2 * v + 1, tm + 1, tr)
        self.t[v] = self.merge(self.t[2 * v], self.t[2 * v + 1])

    def update(self, v, tl, tr, i, x):
        if tl == tr:
            self.t[v] = x
            return
        tm = (tl + tr) >> 1
        if i <= tm:
            self.update(2 * v, tl, tm, i, x)
        else:
            self.update(2 * v + 1, tm + 1, tr, i, x)
        self.t[v] = self.merge(self.t[2 * v], self.t[2 * v + 1])

    def query(self, v, tl, tr, l, r):
        if l > r:
            return 0
        if l == tl and tr == r:
            return self.t[v]
        tm = (tl + tr) >> 1
        return self.merge(
            self.query(2 * v, tl, tm, l, min(tm, r)),
            self.query(2 * v + 1, tm + 1, tr, max(l, tm + 1), r)
        )


N, q = map(int, input().split())
arr = list(map(int, input().split()))

sgt = SegTree(arr)
sgt.build(1, 0, N - 1)

results = []
for _ in range(q):
    ty, a, b = map(int, input().split())

    if ty == 1:  # v, tl, tr, i, x
        sgt.update(1, 0, N - 1, a, b)  # 0-based
        # sgt.update(1, 0, N - 1, a - 1, b)  # 1-based
    else:
        ans = sgt.query(1, 0, N - 1, a, b)  # 0-based
        # ans = sgt.query(1, 0, N - 1, a - 1, b - 1)  # 1-based
        results.append(ans)

for result in results:
    print(result)

File: cpp-templates/seg-tree.cpp
#include <iostream>
#include <vector>
using namespace std;

class SegmentTree {
   private:
    vector<int> arr;
    vector<int> tree;

    int merge(int a, int b) {
        int res = a + b;
        return res;
    }

   public:
    SegmentTree(vector<int> &nums) {
        this->arr = nums;
        this->tree.resize(4 * nums.size(), 0);
    }

    void build(int v, int lo, int hi) {
        if (lo == hi) {
            tree[v] = arr[lo];
            return;
        }
        int mid = lo + (hi - lo) / 2;
        build(2 * v, lo, mid);
        build(2 * v + 1, mid + 1, hi);
        tree[v] = merge(tree[2 * v], tree[2 * v + 1]);
    }

    void update(int v, int lo, int hi, int i, int x) {
        if (lo == hi) {
            tree[v] = x;
            return;
        }
        int mid = lo + (hi - lo) / 2;
        if (i <= mid) {
            update(2 * v, lo, mid, i, x);
        } else {
            update(2 * v + 1, mid + 1, hi, i, x);
        }
        tree[v] = merge(tree[2 * v], tree[2 * v + 1]);
    }

    int query(int v, int lo, int hi, int l, int r) {
        if (l > r) return 0;
        if (lo == l && hi == r) return tree[v];
        int mid = lo + (hi - lo) / 2;
        return merge(query(2 * v, lo, mid, l, min(mid, r)),
                     query(2 * v + 1, mid + 1, hi, max(l, mid + 1), r));
    }
};

void solve() {
    int N, Q;
    cin >> N >> Q;
    vector<int> arr(N);
    for (int &x : arr) cin >> x;

    SegmentTree *ST = new SegmentTree(arr);
    ST->build(1, 0, N - 1);

    vector<int> qs;
    for (int q = 0; q < Q; ++q) {
        int ty, a, b;
        cin >> ty >> a >> b;
        if (ty == 1) {
            // ST.update(1, 0, N - 1, a, b); // 0-based
            ST->update(1, 0, N - 1, a - 1, b);  // 1-based
        } else {
            // int res = ST.update(1, 0, N - 1, a, b); // 0-based
            int res = ST->query(1, 0, N - 1, a - 1, b - 1);  // 1-based
            qs.push_back(res);
        }
    }

    for (int x : qs) cout << x << '\n';
}